fairlearn/metrics/_bootstrap.py

```diff
@@ -0,0 +1,110 @@
+
+"""Utility functions for bootstrap calculations (used for confidence intervals)."""
+
+from warnings import warn
+import pandas as pd
+import numpy as np
+from scipy.stats import norm
+import numbers
+
+def process_ci_bounds(ci, ci_method, precision=6):
+    '''Parses user inputs to construct confidence intervals from bootstrap.
+       This function interprets integer and float inputs as the desired
+       width of a two-sided confidence interval, and lists and tuples as
+       exact quantiles to retrieve.
+
+       Parameters
+        ----------
+        ci : float, list, tuple
+        Determines the quantiles reported for confidence intervals. All entries must
+        be in the range [0, 1]. A single float is interpreted as the width of the 
+        desired confidence interval (e.g. 0.95 is a 95% interval), while a List or Tuple
+        reports exact quantiles from the empirical bootstrap distribution 
+        (e.g. (0, 0.33, 0.5, 0.66, 1) will calculate the min, 33rd, 50th, 66th, and max quantiles).
+        
+        method: str
+        One of {'percentile', 'bias-corrected'}. Used to validate CI inputs.
+
+        precision: int
+        Number of decimals of precision to round quantile calculations to.
+    '''
+    if ci_method not in ['percentile', 'bias-corrected']:
+        raise ValueError(f"Parameter ci_method: {ci_method} must be set to 'percentile' or 'bias-corrected'.")
+
+    if isinstance(ci, (int, float)) and (0 <= ci <= 1):
+        return tuple(np.round((0.5 - ci/2, 0.5 + ci/2), decimals=precision))
+
+    elif isinstance(ci, (tuple, list)):
+        if all(((0 <= x <= 1) for x in ci)):
+            return tuple(ci)
+        else:
+            raise ValueError(f"All desired confidence interval bounds must be within [0, 1].")
+
+    else:
+        raise ValueError(f"Input to 'ci' of type {type(ci)} must be of type int, float, tuple, or list.")
+    
+
+def create_ci_output(bootstrap_runs, ci, sample_estimate, interval_type='percentile'):
+    '''Calculates and formats confidence intervals from bootstrap samples.'''
+    
+    #--- Check if there is significant bias in sample statistics ---
+    bootstrap_mean = np.mean(bootstrap_runs, axis=0)
+
+    # Need to use this incredibly odd workaround to support frames with "object" dtypes
+    # object typed frames are produced by default from .__group()
+    bootstrap_std = np.sqrt(np.var(bootstrap_runs, axis=0, dtype='float64'))
+    bias = bootstrap_mean - sample_estimate
+    bias_adjusted_mean_estimate = 2*sample_estimate - bootstrap_mean # Might be useful to publish?
+
+    # Efron and Tibshirani (1993) guidance on identifying "large" bias via bootstrap
+    high_bias_regions = np.abs(bias) >= (0.25 * bootstrap_std)
+    if np.any(high_bias_regions):
+        #TODO: Add bias indication regions into warning. Simply printing all "high_bias_regions" is overwhelming.
+        warn(
+            f"""Some statistics calculated by MetricFrame are detected to have high bias 
```

Comment:
Please be more definite in the error message - you know which metrics came back with large errors, so tell the user which ones.... is that what you're getting at in the TODO?

Reply:
> is that what you're getting at in the TODO?

Yes exactly -- I was struggling a bit with how to format the warning properly without creating a (potentially) overwhelmingly large message. We're computing so many different disaggregated metrics -- we calculate bias in each cell of .overall, .by_group, difference to overall/between groups, and ratio to overall/between groups -- and any one of the cells in any of the frames might have high statistical bias. Actionable information for a user would include: 

- Name of the frame (.overall vs. difference_between_groups, etc.)
- Name of the metric
- index values (control features and sensitive attributes that make up the calculation)

and there might be potentially dozens of these. Wasn't sure if I should try to format all this info line by line, which could lead to a large warning message, or throw a generic warning and capture all of the details in a property that a user could inspect later

Reply:
Understood - I hadn't realised just how large the list of potential 'hits' could be. I can see that extra discussion is warranted around that. Could you format them into a list, attach it as an extra field to the object, and tell users to inspect it?